contracts/val-ref/src/contract.rs

```diff
@@ -0,0 +1,194 @@
+        validatorset::{ExecuteMsg as ValidatorSetExecuteMsg, OnTopEditOperation},
+    },
+    state::val_ref::{CORE_ADDRESS, REFS, VALIDATORS_SET_ADDRESS},
+};
+use neutron_sdk::bindings::{msg::NeutronMsg, query::NeutronQuery};
+
+pub const CONTRACT_NAME: &str = env!("CARGO_PKG_NAME");
+pub const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+#[cfg_attr(not(feature = "library"), cosmwasm_std::entry_point)]
+pub fn instantiate(
+    deps: DepsMut<NeutronQuery>,
+    _env: Env,
+    _info: MessageInfo,
+    msg: InstantiateMsg,
+) -> ContractResult<Response<NeutronMsg>> {
+    cw2::set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+    cw_ownable::initialize_owner(deps.storage, deps.api, Some(&msg.owner))?;
+
+    let core = deps.api.addr_validate(&msg.core_address)?;
+    CORE_ADDRESS.save(deps.storage, &core)?;
+
+    let validators_set = deps.api.addr_validate(&msg.validators_set_address)?;
+    VALIDATORS_SET_ADDRESS.save(deps.storage, &validators_set)?;
+
+    Ok(response(
+        "instantiate",
+        CONTRACT_NAME,
+        [
+            attr("core_address", core),
+            attr("validators_set_address", validators_set),
+        ],
+    ))
+}
+
+#[cfg_attr(not(feature = "library"), cosmwasm_std::entry_point)]
+pub fn execute(
+    deps: DepsMut<NeutronQuery>,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> ContractResult<Response<NeutronMsg>> {
+    match msg {
+        ExecuteMsg::UpdateOwnership(action) => {
+            cw_ownable::update_ownership(deps.into_empty(), &env.block, &info.sender, action)?;
+            Ok(response::<(&str, &str), _>(
+                "execute-update-ownership",
+                CONTRACT_NAME,
+                [],
+            ))
+        }
+        ExecuteMsg::BondCallback(bond_hook) => execute_bond_hook(deps, info, bond_hook),
+        ExecuteMsg::UpdateConfig {
+            core_address,
+            validators_set_address,
+        } => execute_update_config(deps, info, core_address, validators_set_address),
+        ExecuteMsg::SetRefs { refs } => execute_set_refs(deps, info, refs),
+    }
+}
+
+fn execute_bond_hook(
+    deps: DepsMut<NeutronQuery>,
+    info: MessageInfo,
+    bond_hook: BondHook,
+) -> ContractResult<Response<NeutronMsg>> {
+    let core = CORE_ADDRESS.load(deps.storage)?;
+    if info.sender != core {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    let mut messages = vec![];
+    let mut attrs = vec![];
+    if let Some(r#ref) = bond_hook.r#ref {
+        attrs.push(attr("ref", &r#ref));
+        if let Some(validator_address) = REFS.may_load(deps.storage, &r#ref)? {
+            attrs.push(attr("validator", &validator_address));
+            let exchange_rate: Decimal = deps.querier.query_wasm_smart(
+                core,
+                &drop_staking_base::msg::core::QueryMsg::ExchangeRate {},
+            )?;
+            let on_top_increase = bond_hook.dasset_minted * exchange_rate;
```

Comment:
why don't you want to use pure `amount` instead of `dasset_minted * exchange_rate`, you already know amount when you are calling this hook.

Reply:
because the field `amount` contains the amount of an incoming denomination, which might be an LSM share. what I need in this context is strictly the amount of ATOMs (or other original assets like TIA or OSMO)